core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java

```diff
@@ -0,0 +1,181 @@
+  private boolean headerWritten;
+  private boolean finished;
+  private Optional<Integer> footerSize = Optional.empty();
+  private Optional<Long> fileSize = Optional.empty();
+
+  PuffinWriter(
+      OutputFile outputFile, Map<String, String> properties, boolean compressFooter,
+      PuffinCompressionCodec defaultBlobCompression) {
+    Preconditions.checkNotNull(outputFile, "outputFile is null");
+    Preconditions.checkNotNull(properties, "properties is null");
+    Preconditions.checkNotNull(defaultBlobCompression, "defaultBlobCompression is null");
+    this.outputStream = outputFile.create();
+    this.properties = ImmutableMap.copyOf(properties);
+    this.footerCompression = compressFooter ? PuffinFormat.FOOTER_COMPRESSION_CODEC : PuffinCompressionCodec.NONE;
+    this.defaultBlobCompression = defaultBlobCompression;
+  }
+
+  @Override
+  public void add(Blob blob) {
+    Preconditions.checkNotNull(blob, "blob is null");
+    checkNotFinished();
+    try {
+      writeHeaderIfNeeded();
+      long fileOffset = outputStream.getPos();
+      PuffinCompressionCodec codec = MoreObjects.firstNonNull(blob.requestedCompression(), defaultBlobCompression);
+      ByteBuffer rawData = PuffinFormat.compress(codec, blob.blobData());
+      int length = rawData.remaining();
+      writeFully(rawData);
+      writtenBlobsMetadata.add(new BlobMetadata(blob.type(), blob.inputFields(), fileOffset, length,
+          codec.codecName(), blob.properties()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public Metrics metrics() {
+    return new Metrics();
+  }
+
+  @Override
+  public long length() {
+    return fileSize();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!finished) {
+      finish();
+    }
+
+    outputStream.close();
+  }
+
+  private void writeHeaderIfNeeded() throws IOException {
+    if (headerWritten) {
+      return;
+    }
+
+    this.outputStream.write(MAGIC);
+    this.headerWritten = true;
+  }
+
+  public void finish() throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
+    Preconditions.checkState(!footerSize.isPresent(), "footerSize already set");
+    long footerOffset = outputStream.getPos();
+    writeFooter();
+    this.footerSize = Optional.of(Math.toIntExact(outputStream.getPos() - footerOffset));
+    this.fileSize = Optional.of(outputStream.getPos());
+    this.finished = true;
+  }
+
+  private void writeFooter() throws IOException {
+    FileMetadata fileMetadata = new FileMetadata(writtenBlobsMetadata, properties);
+    ByteBuffer footerJson = ByteBuffer.wrap(
+        FileMetadataParser.toJson(fileMetadata, false).getBytes(StandardCharsets.UTF_8));
+    ByteBuffer footerPayload = PuffinFormat.compress(footerCompression, footerJson);
+    outputStream.write(MAGIC);
+    int footerPayloadLength = footerPayload.remaining();
+    writeFully(footerPayload);
```

Comment:
Minor: why not call `IOUtil.writeFully(outputStream, footerPayload)` directly rather than having a private `writeFully` method that just adds the output stream?

Reply:
https://github.com/apache/iceberg/pull/5088